Refactor eventdata - work in progress

or opened the pull request:
This is far from complete, the focus is unit testing and getting the main functions covered.
- deprecated models (`Event`, `EventData`, `EventInstance`, `EventInstanceData`) were stripped of their logic and left in only to keep the rest of the code happy for now, planned to be removed in the commit after data migration
- implemented `NewEvent`, which will replace `EventInstance` and `EventInstanceData` and ultimately be renamed to `Event`, representing _actual_ events
- implemented `EmarsysEventId` as a mapping of event name to Emarsys ID, this will be synced on the fly when necessary (i.e. when there is no ID for a given event name yet)

Rough idea:
- settings.EMARSYS_EVENTS configures all events and their expected data
- `django_emarsys.util.validate_configuration` sanity checks all `EMARSYS_*` settings and should be called at the end of settings or perhaps somewhere in the build process, `tests/configuration.py` tests this functionality
- `service.trigger_event` is exposed to be used by the parent code, it maps almost exactly to the former `Event.objects.trigger()`. `tests/trigger_event.py` tests this functionality

Open questions:
- [x] Is logging (error for things that will break the code, warning for things that are not perfect but not a problem) in `validate_configuration` enough or should it raise an exception if at least one error occurred?
  - made it a system check
- [x] Can the code assume `EMARSYS_*` settings are correct and use them as expected, as long as errors are caught and logged into `NewEvent` objects?
  - Yes.

Issues:
- [x] It's probably worth pondering what code belongs in Managers and what in Models and (what currently still is called) Service calls.
- [x] `NewEvent.set_data` might not be the best solution yet, because it does quite a bit, in particular things that can break, if `EMARSYS_*` settings are invalid.
  - Using `set_parameter()` now for each parameter, the validation is done outside of the Model and `set_parameter()` only has to decide how it will store the object.
- [x] Having to import `EmarsysEventId` inside `NewEventManager` to avoid circular dependencies is a bit awkward. Perhaps there's a better way.
  - The Emarsys event ID must now be passed along to `NewEventManager`, so the calling code has to interact with all the managers involved, but neither of the managers has to interact with another manager or model.

Testing:

```
$ py.test
```

Ideally for prettiness:

```
$ pip install git+https://github.com/or/pytest-sugar
```

To-Dos:
- [x] delete stale Emarsys-IDs
- [x] remove async stub
- [x] reconsider warning levels (e.g. "sync needed")
- [x] ~~~use different `JSONField`~~~ (this would bind `django-emarsys` to a PostgreSQL backend)
- [x] use state constants instead of string literals
- [x] list/set comprehensions instead of loops
- [x] `pk` instead of `id`
- [x] plainer log messages (e.g. missing settings)
- [x] `str.rsplit` instead of `str.split`
- [x] delete `urls.py`
- [x] `assertSetEqual`
- [x] string continuation in parentheses
- [x] try `util.log = Mock()`
- [x] simplify configuration validation
- [x] separate `validate_configuration` and `validate_settings` for easier testing

The lib now behaves like before, last steps:
- [x] refactor EventParam
- [x] move code around for great justice
- [x] remove `User` foreign key in `NewEventInstance` in favour of treating it like any other event parameter
- [x] include state and error message, if any, in `NewEventInstance` view
- [x] ~~~include `email_id` or other return values from api in `NewEventInstance`~~~
  The API doesn't provide anything of the sort: `u'{"replyCode":0,"replyText":"OK","data":{}}'`
- [x] migrate old `EventInstance`s to `NewEventInstance`
- [x] delete old models
- [x] rename `NewEvent` to `Event` and `NewEventInstance` to `EventInstance`
- [x] get `Event` and `EventData` data from production database to get the latest version, use that in `settings.EMARSYS_EVENTS`

Another round:
- [x] review adjustments
- [x] refactor `trigger_automatic_event()` and `trigger_manual_event()` to `trigger_event()` and `trigger_event(..., manual=True)`
- [x] make `sync_events` ignore unconfigured event names

Last tweaks:
- [x] `requirements.txt`
- [x] `test_helper.py`

mbertheau commented:
General direction seems fine to me.

I'll make some comments anyway - none of them imply that you're not already aware of what they're pointing out. However, some of them point out issues that can complicate things when you finalize this, so I want to get these out early.

mbertheau commented on django_emarsys/managers.py:
```diff
@@ -2,81 +2,94 @@
+                defaults={'emarsys_id': emarsys_id})
+            if not created:
+                obj.emarsys_id = emarsys_id
+                obj.save()
```
Need to delete events that were deleted on emarsys side.

or replied:
Done.

mbertheau commented on django_emarsys/managers.py:
```diff
@@ -2,81 +2,94 @@
+    def trigger(self, event_name, recipient_email, user=None, data=None,
+                source='automatic', async=False):
+        if async:
+            raise NotImplementedError
```
Kind of a stupid stub `async` that I put there. Ok to take out.

or replied:
Ok, I'll take it out for now, but I think it's a very good and important idea to do this asynchronously... especially if it might need to retrieve the ID now and then (which in practice is ultra rare, of course).

mbertheau commented on django_emarsys/managers.py:
```diff
@@ -2,81 +2,94 @@
+        event = self.create(
+            event_name=event_name,
+            recipient_email=recipient_email,
+            user=user,
```
User is a special snowflake. The only reason it's not a kwarg like the others is so that I can have a user fk column in the old eventinstance table. However, only by convention user is set to the recipient for events where the recipient is a django user. Just fyi.

mbertheau commented on django_emarsys/managers.py:
```diff
@@ -2,81 +2,94 @@
+                raise UnknownEventNameError(event_name)
+
+            event.set_data(data)
+        except Exception as e:
```
catch too general.

mbertheau commented on django_emarsys/managers.py:
```diff
@@ -2,81 +2,94 @@
+            event_id = EmarsysEventId.objects.get(event_name=event_name)
+        except EmarsysEventId.DoesNotExist:
+            log.warning("Emarsys-ID for event '{event}' unknown, "
+                        "syncing...".format(event=event_name))
```
Warning might be too strong here. If syncing gets the event, that's a normal and expected case, i.e. info or even debug. If the event isn't there even after syncing, you get a stronger log later (I believe? you should.).

or replied:
+1

mbertheau commented on django_emarsys/models.py:
```diff
@@ -141,12 +76,63 @@ class EventInstance(models.Model):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True)
+    recipient_email = models.CharField(max_length=1024)
+    context = JSONField(null=True)
+    data = JSONField()
```
In machtfit-oscar we use postgres.fields.JSONField which is API-compatible with https://docs.djangoproject.com/en/1.8/ref/contrib/postgres/fields/#arrayfield

I don't know about jsonfield's API but we should use the field that's compatible with the upcoming https://docs.djangoproject.com/en/dev/ref/contrib/postgres/fields/#jsonfield

or replied:
Ah, ok, jsonfield is the one I'm most familiar with. I should've checked the rest of the code first.

or replied:
For the record: we decided to use `jsonfield`, because `postgres.fields.JSONField` is PostgreSQL specific.

mbertheau replied:
That triggers a thought: you mentioned, and I agree, that often, when you squash commits to create a clean history, you lose the final rationale for why certain design decisions were reconsidered and changed in the process.

I think this information would be best held in the new commit messages. I.e. in the commit that introduces the jsonfield, mention that we considered the other thing because of this, but then because of that we used the one thing anyway.

mbertheau commented on django_emarsys/models.py:
```diff
@@ -141,12 +76,63 @@ class EventInstance(models.Model):
+    result = models.CharField(max_length=1024, blank=True)
+    result_code = models.CharField(max_length=1024, blank=True)
+    state = models.CharField(max_length=1024, choices=STATE_CHOICES,
+                             default='sending')
```
Should use python symbols (i.e. "constants") instead of string literals to catch typos at compile time.

or replied:
+1

mbertheau commented:
I may also at times comment things that I wrote myself. It's never too late to review code that you've written yourself! :)

mbertheau commented on django_emarsys/models.py:
```diff
@@ -162,16 +148,39 @@ def label(self):
-        unique_together = [('event_trigger', 'event_data')]
+        actual_args = set()
+        for key, value in data.items():
+            actual_args.add(key)
```
A bit shorter: `actual_args = set(data)`

mbertheau commented on django_emarsys/models.py:
```diff
@@ -162,16 +148,39 @@ def label(self):
-    content_object = GenericForeignKey('content_type', 'object_id')
+        expected_args = set()
+        for event_data in events[self.event_name]:
+            expected_args.add(event_data['arg'])
```
`expected_args = {event_data['arg'] for event_data in events[self.event_name]}`

mbertheau commented on django_emarsys/models.py:
```diff
@@ -162,16 +148,39 @@ def label(self):
+                                 .format(model=model_name, arg=arg,
+                                         value=data[arg]))
+
+            stored_data[arg] = (model_name, data[arg].id)
```
`pk` should give you the PK in cases where you don't have an `id` column.

mbertheau commented on django_emarsys/service.py:
```diff
@@ -0,0 +1,25 @@
+from django_emarsys import api
```
Convention in this repo is to start files with

```
# -*- coding: utf-8 -*-

from __future__ import unicode_literals
```

mbertheau replied:
`service.py` -> `event.py`

or replied:
I'm not so sure about that one. `service` might not be the best choice, but to name the module after the model seems confusing.

I'll keep this in mind.

mbertheau replied:
The model is named after the concept, and this file contains the primary API for the concept. :)

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,97 @@
+
+def validate_configuration():
+    if not getattr(settings, 'EMARSYS_ACCOUNT', None):
+        log.warning("no EMARSYS_ACCOUNT setting set")
```
EMARSYS_ACCOUNT not set

or replied:
It was a George R. R. Martin quote, but ok!

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,97 @@
+def get_model(model_string):
+    chunks = model_string.split('.')
+    app_label = '.'.join(chunks[:-1])
+    model_name = chunks[-1]
```
Use https://docs.python.org/2/library/stdtypes.html#str.rsplit

or replied:
Noice.

mbertheau commented on sandbox/urls.py:
```diff
@@ -1,10 +1,7 @@
-    # url(r'^blog/', include('blog.urls')),
-
+urlpatterns = patterns(
+    '',
```
Unnecessary change. This is just the default file created by Django's startproject and it's not used otherwise.

or replied:
flake8 conformity, but I'll delete it instead.

mbertheau commented on tests/configuration.py:
```diff
@@ -0,0 +1,105 @@
+            if symbol.startswith('EMARSYS_'):
+                delattr(settings, symbol)
+
+        util.validate_configuration()
```
Simpler to just pass a settings dict to `validate_configuration`.

or replied:
How do you mean?

mbertheau replied:
Decouple `validate_configuration` from the global django settings by changing its signature to accept a settings dict and passing `django.conf.settings` when using it. In test you can make up a settings dict on the fly without having to mock settings or deleting attributes on a module (which isn't something bad per se, just that creating a simple dict is simpler). That's an example of how decoupling increases testability.

or replied:
Still don't get where it becomes a dict. It'd still be a Settings object? Is there an elegant way to create a Settings object on the fly?
Or do you mean that it wouldn't matter as long as `validate_configuration` only uses `hasattr()` and `getattr()`?

mbertheau replied:
Ah, right, I neglected that.

So you'd have to call `validate_configuration` with
`{key: getattr(settings, name) for name in dir(settings)}` which is not the most beautiful code, but a fair price to pay for the decoupled function. I can imagine something like this:

``` python
def validate_settings(settings):
    return validate_configuration({key: getattr(settings, name) for name in dir(settings)})

def validate_configuration(conf):
    ...
```

And then you test `validate_configuration`, which has the guts that should be tested.

mbertheau commented on tests/configuration.py:
```diff
@@ -0,0 +1,105 @@
+        if expected_messages:
+            raise self.failureException(
+                "expected log message(s) missing: {}"
+                .format(list(expected_messages)))
```
assertEqual does all that for you by calling https://docs.python.org/3.5/library/unittest.html#unittest.TestCase.assertSetEqual . (Django includes unittest2, which is unittest from Python 3)

or replied:
Ah, of course. That' better.

mbertheau commented on tests/configuration.py:
```diff
@@ -0,0 +1,105 @@
+            "missing data key for event 'test event 2': 'arg'",
+            "missing data key for event 'test event 2': 'model'",
+            "bad model 'invalid_app.User' for event 'test event 3': "
+            + "No installed app with label 'invalid_app'.",
```
`+` is unnecessary here. To make it more clear visually that this is actually a string continuation, consider putting the string in parentheses.

or replied:
Solely because of flake8. But parentheses are accepted and allow one space extra indentation for the continuation, I'll use that.

mbertheau commented on django_emarsys/test_helper.py:
```diff
@@ -63,3 +64,19 @@ def assertEmarsysEvents(self, events, all=True):
+        self.messages.append(record)
+
+    def reset(self):
+        self.messages = []
```
Can replace this by setting `util.log = Mock()` in the test, which then collects all method calls itself.

or replied:
I'll try that out.

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,97 @@
+                              .format(event=event, model=model))
+                else:
+                    try:
+                        get_model(model)
```
`get_model` is a notorious bitch and can be used only after the models were loaded successfully. This limits the places where a call to this function can be placed. Just fyi.

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,97 @@
+                if key not in ('name', 'arg', 'model'):
+                    log.warning(
+                        "unexpected data key for event '{event}': '{key}'"
+                        .format(event=event, key=key))
```
I believe this code can be written simpler. For example, at the data level, by comparing `{'name', 'arg', 'model'}` with `set(data)` you can test for missing and additional keys. Then at the event level, you can look at all names `{data['name'] for`

I just saw `for data in data`. Confusing.

I'll pretend it says `for datom in data` henceforth.

At the event level, you can look at all names `[datom['name'] for datom in data if 'name' in datom]` and handle invalid and duplicate names. Similar for args.

or replied:
Will look into it.

or replied:
I think that'd shorten some of the code, but result in less specific error/warning messages.
`{'name', 'arg', 'model'} != set(data)` would have to look into it in detail again or just output "there's a mismatch: <expected>, <actual>", which is less helpful.

Also, an additional keyword is no problem, so it's only a warning, but one missing is an error. 

Similarly, finding which `arg` or `name` is duplicated with the list you gave seems more complicated in the end, while I agree that the check "is anything duplicated?" is shorter then.

So if the same level of reporting detail is wished for (I wish for it!), then the code probably would turn out complicated again.

But I think there are some lines where `set` arithmetic will make the code less bulky and I'll try to reduce the boiler plate size.

mbertheau commented:
1st question: Settings validation should probably happen in a check https://docs.djangoproject.com/en/1.7/topics/checks/

mbertheau commented:
2nd question: Yes; also `trigger` should give the caller the option to detect success or failure of the event.

mbertheau commented:
1st issue: I'd put the minimal amount of code in managers and models that you can get away with. Reason is that any code in there is tightly coupled to the manager, model and the django database, which makes it harder to test. That's my general opinion since a few months, and I adhered to it in the finance app - I'm encouraging you to look at it. However in the finance app I made the error to put a lot of state in the Invoice and Settlement objects, which caused a number of subtle bugs. I hope to have found most of them. However, my stance now is that state is bad for testability, composability, reusability and requires you to follow the whole program flow to make assertions about its effects. Now I prefer a more functional style where I work with pure functions. That means it has no side-effects and their return value depends only on its argument values. The I/O (side-effects) should be bundled in one place.

mbertheau commented:
2nd issue: I agree. the data should be validated outside of `set_data`, before the event is created.

mbertheau commented:
3rd issue: I agree. Looking at where `self` is used in the methods of `NewEventManager` - all that code can be in a plain function in `event.py`. That eliminates the problem, which is one of coupling things that shouldn't be coupled.

mbertheau commented:
Test instructions may be incomplete, I get:

```
~/src/django-emarsys $ py.test
============================================================================================================================================================================ test session starts ============================================================================================================================================================================
platform linux2 -- Python 2.7.6 -- py-1.4.30 -- pytest-2.7.2
rootdir: /home/markus/src/django-emarsys, inifile: 
plugins: django
collected 0 items / 1 errors 

================================================================================================================================================================================== ERRORS ===================================================================================================================================================================================
______________________________________________________________________________________________________________________________________________________________ ERROR collecting django_emarsys/test_helper.py _______________________________________________________________________________________________________________________________________________________________
django_emarsys/test_helper.py:9: in <module>
    from django_emarsys.models import EventInstance
django_emarsys/models.py:10: in <module>
    from django.contrib.contenttypes.fields import GenericForeignKey
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/contrib/contenttypes/fields.py:14: in <module>
    from django.contrib.contenttypes.models import ContentType
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/contrib/contenttypes/models.py:140: in <module>
    class ContentType(models.Model):
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/contrib/contenttypes/models.py:141: in ContentType
    name = models.CharField(max_length=100)
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/db/models/fields/__init__.py:1012: in __init__
    super(CharField, self).__init__(*args, **kwargs)
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/db/models/fields/__init__.py:146: in __init__
    self.db_tablespace = db_tablespace or settings.DEFAULT_INDEX_TABLESPACE
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/conf/__init__.py:46: in __getattr__
    self._setup(name)
../../.virtualenvs/django-emarsys/local/lib/python2.7/site-packages/django/conf/__init__.py:40: in _setup
    % (desc, ENVIRONMENT_VARIABLE))
E   ImproperlyConfigured: Requested setting DEFAULT_INDEX_TABLESPACE, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
========================================================================================================================================================================== 1 error in 0.24 seconds ==========================================================================================================================================================================
```

or commented:
Aye, `pytest.ini` was missing. Should work now.

mbertheau commented:
Ack, works now. pytest seems neat. We should see later if it gives us less hassle than nose in machtfit-oscar :)

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,112 @@
+        app_label = None
+        model_name = model_string
+
+    return apps.get_model(app_label=app_label, model_name=model_name)
```
Turns out `get_model` accepts a model_string directly: https://docs.djangoproject.com/en/1.7/ref/applications/#django.apps.apps.get_model

mbertheau replied:
What does `get_model(None, ...)` do anyway?

or replied:
Nice, I should read more than the signature sometimes.

It results in an error. The question is how to handle that case in that function. It cannot recuperate anyway.
Using the `apps.get_model(model_string)` directly will be easier, but it throws `ValueError("need more than 1 value to unpack")`. Oh well.

mbertheau commented on django_emarsys/util.py:
```diff
@@ -0,0 +1,112 @@
+def validate_settings():
+    validate_configuration({key: getattr(settings, key)
+                            for key in dir(settings)
+                            if hasattr(settings, key)})
```
Why the `hasattr`? Does `dir` return something for which `hasattr` doesn't hold?

or replied:
Surprisingly yes. This happened when I still tweaked settings in the tests, so it may be an issue with overridden settings, but I included it just to be safe.

I'll comment on it in the doc string.

mbertheau replied:
Please include the specific circumstances in which the situation occurrs, that makes checking with `hasattr` necessary.

or replied:
```
    @override_settings()
    def test_strange_dir_issue(self):
        delattr(settings, 'EMARSYS_ACCOUNT')
        self.assertTrue('EMARSYS_ACCOUNT' in dir(settings))
        self.assertFalse(hasattr(settings, 'EMARSYS_ACCOUNT'))
```

Only if the setting exists in the original settings and then is removed. But since `validate_settings()` might experience more testing, I'll leave the check in and describe it as you said.

mbertheau commented on tests/configuration_tests.py:
```diff
@@ -0,0 +1,119 @@
+                           if hasattr(settings, key)}
+        util.validate_settings()
+
+        mock_validate_configuration.assert_called_with(expected_config)
```
What's the point of this test?

The benefit of a test is correlated with the probability that the test catches a bug in the implementation, and inversely correlated with my confidence that the tested code works (without testing it). (That's why simpler code is better - I can be more confident that it works correctly if it's easier to reason about it.)

The cost of a test is that it has to be maintained, adds running time to the test suite, and may be buggy itself.

In this case the test code duplicates the function's implementation. That reduces the benefit to zero - if I have a bug in the implementation, it will be in the test as well.

mbertheau replied:
(Oliver answered:)

Yeah, I was unsure about the replication of the code there, but it's not entirely trivial and I want to make sure `validate_configuration()` is called correctly. It also makes sure `validate_settings()` is called at least once in tests.

Future changes to `validate_settings()` could break this test and in that case it's worth a second look at the change, I think the test is justified.

With currently 7 tests the question of test running time and maintenance is a non-issue. Let's err on the side of having a few tests too many instead of missing them.

mbertheau replied:
Ok, I'll try to understand your arguments:

"it's not entirely trivial":

What is?

"I want to make sure that validate_configuration is called correctly"

I'm 100% sure that that's the case by just looking at the code. That part is outright trivial.

"It also makes sure `validate_settings()` is called at least once in tests."

If this is about coverage: coverage has diminishing returns. To achive the last 30 or 20 or 5%, you end up testing trivial code, where you expend an effort to keep the test happy that's unproportionally higher than the probability that the test will catch a bug.

"in that case it's worth a second look at the change"

Do you mean that you have to make the change in two places, and that's the second look? If so, the second look is provided by a code review.

I agree time running time is insignificant in this particular case.

"maintenance is a non-issue"

How so?

(Feel free to postpone answering to a later date.)

or committed:
Use a quasi-singleton client object in api

That way no code is executed when importing the module. In particular
code that is using settings attributes that may not have been checked
yet.

mbertheau commented on django_emarsys/checks.py:
```diff
@@ -0,0 +1,134 @@
+                   Error: log.error,
+                   Critical: log.critical}[type(message)]
+
+    log_message(message.msg)
```
Hmm, I think if it made sense to log failing checks, Django would already do that. What's the rationale? I remember you mentioned that, but I don't remember the rationale.

I'd rather not log. Django isn't going to start up with failing checks anyway.

or replied:
Alright, I'm convinced. I left them out.

mbertheau commented on django_emarsys/checks.py:
```diff
@@ -0,0 +1,134 @@
+def validate_configuration(config):
+    messages = []
+    if not config.get('EMARSYS_ACCOUNT'):
+        add_message(messages, Critical("EMARSYS_ACCOUNT not set"))
```
I'd rather have a variable be modified in one place, than in two. As it stands, add_message has side effects.

Since this point is moot if we decide not to log the messages I'll wait until we make that decision.

mbertheau commented on django_emarsys/checks.py:
```diff
@@ -0,0 +1,134 @@
+                messages,
+                Error("bad model '{value}' for event "
+                      "'{event}': {error}"
+                      .format(value=value, event=event, error=e)))
```
I find it really hard to follow what this code does. What you essentially have to do is pretend you're a computer and follow the program flow from line 75 through all the loops. I'll try to write something that I believe is easier to read and understand.

mbertheau replied:
For example, I just arrived at Line 110. To figure out that the meaning is "If this key should have unique values" took me too long.

or replied:
Aye, the duplicate code probably doesn't belong in `validate_event_param`, actually.

I think the rest does a straight-forward check for everything that needs to be checked about a parameter key. But I'll wait for your approach! :)

mbertheau replied:
https://github.com/machtfit/django-emarsys/pull/7

mbertheau commented on django_emarsys/checks.py:
```diff
@@ -0,0 +1,134 @@
+        add_message(
+            messages,
+            Error("invalid parameter {key} for event '{event}': '{value}'"
+                  .format(key=key, value=value, event=event)))
```
Also this error message is not so helpful because it doesn't say which parameter has the invalid key. You can probably derive that easily from the key value, but nevertheless.

or replied:
Yeah, I thought about this, but there is no "id" or anything. Worst case is that the key is the only one present.
It might be useful to output the whole `dict`.

or replied:
We could also make sure `arg` is unique by making the event data another `dict` and maybe go for a tuple then:

``` python
EMARSYS_EVENTS = {
  'event name': {
    'user': ('Benutzer', 'auth.User'),
    'product': ('Angebot', 'catalogue.Product'),
  },
  [...]
}
```

This might be a more natural configuration.

mbertheau replied:
I see now that's hard to do because you don't know which keys you have for the parameter.

mbertheau replied:
I like the nested dicts.

mbertheau commented on django_emarsys/checks.py:
```diff
@@ -0,0 +1,134 @@
+    VALID_VALUES = {
+        'name': re.compile(r'^[a-zA-Z0-9_ ]+$'),
+        'arg': re.compile(r'^[a-z][a-z0-9_]*$'),
+        'model': re.compile(r'^[a-z][a-z0-9_]*[.][A-Z][a-zA-Z0-9_]*$'),
```
I'll skip this regex because get_model will be more accurate.

mbertheau commented on tests/configuration_tests.py:
```diff
@@ -0,0 +1,105 @@
+    True
+
+    """
+    return {(m.level, m.msg) for m in messages}
```
That message could be used nicely in a `assertCheckMessagesEqual` or similar.

mbertheau replied:
s/That message/That code/

mbertheau commented on django_emarsys/config.py:
```diff
@@ -0,0 +1,160 @@
+
+    messages += flatten([_validate_event(event, params)
+                         for event, params
+                         in events.items()])
```
Can probably get rid of the `[]` here - flatten should iterate over a generator expression just fine, like in line 126 in `_validate_event`.

mbertheau commented on django_emarsys/config.py:
```diff
@@ -0,0 +1,160 @@
+    if not isinstance(param, tuple) or len(param) != 2:
+        messages.append(
+            Error("invalid parameter definition '{event}': '{arg}' => {param}"
+                  .format(param=param, arg=arg, event=event)))
```
Maybe we can move this check into `_validate_event` like this:

``` python
try:
    name, model = param
except ...:
    ...
```

And then work with `name` and `model` instead of `param[0]` and `param[1]`.

or replied:
That'd destroy the list comprehension where `_validate_event_param()` is called.

I think this is really quite tidy already and it makes sense that `_validate_event()` does NOT have to know about `param`'s structure.

or replied:
Slight correction: I guess it needs to know about it for the duplicate detection of `name`.

Oh, and that also means the duplicate check will fail if one param is malformed... damn. That's one reason why I tried to test those things iteratively, mind.

or replied:
Fixed that and added a test for it.

mbertheau replied:
If you want to check for duplicate names, that needs to happen on the level of the event. Ergo on that level you need to know about params structure.

You can write the comprehension like this:

``` python
return messages + flatten(
    _validate_event_param(event, arg, name, model)
    for arg, (name, model) in params.items())
```

But of course that still leaves us with `param[0]` above that and destructuring the tuple outside the comprehension probably complicates things. Hrm. Maybe a namedtuple is not so bad after all? :)

or replied:
The latest version catches that case. Still uses param[0], but this will be explained in doc strings describing the two functions. I think we've converged on good code (if not perfect).

or commented on django_emarsys/apps.py:
```diff
@@ -0,0 +1,14 @@
+    def ready(self):
+        @register()
+        def check_settings(app_configs):
+            return validate_settings()
```
Registering checks in `AppConfig.ready()` is indeed the recommended way.

Unfortunately, `ready()` also is called for tests, so let's just accept that even the test run will validate the given settings.

mbertheau commented on django_emarsys/apps.py:
```diff
@@ -0,0 +1,27 @@
+                # apparently Django's check framework can't handle
+                # unicode correctly, so replace all non-ascii characters
+                # with '?'
+                message.msg = re.sub(r'[^\x00-\x7F]+', '?', message.msg)
```
`.encode('ascii', 'replace')` is simpler.

But I don't buy that django's check framework can't handle unicode correctly.

I suspect that that's a python 2 issue which is convinced stdout can only understand ASCII. This can be worked around like so: https://github.com/machtfit/machtfit-oscar/blob/43d1fc7af86db67aaa1afee61f2edc1721f18ee9/apps/catalogue/management/commands/generate_products.py#L42

At which point is there actually a character in the message that's not in ASCII?

or replied:
Ah, nice. Had forgotten about `'replace'`.

Event names and parameter names can and do contain non-ASCII characters ("Email bestätigen", etc.).

It has nothing to do with stdout:

```
Performing system checks...

Unhandled exception in thread started by <function wrapper at 0x7fadb1a35500>
Traceback (most recent call last):
  File "/home/thi/.virtualenvs/machtfit-oscar/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 222, in wrapper
    fn(*args, **kwargs)
  File "/home/thi/.virtualenvs/machtfit-oscar/local/lib/python2.7/site-packages/django/core/management/commands/runserver.py", line 105, in inner_run
    self.validate(display_num_errors=True)
  File "/home/thi/.virtualenvs/machtfit-oscar/local/lib/python2.7/site-packages/django/core/management/base.py", line 362, in validate
    return self.check(app_configs=app_configs, display_num_errors=display_num_errors)
  File "/home/thi/.virtualenvs/machtfit-oscar/local/lib/python2.7/site-packages/django/core/management/base.py", line 399, in check
    formatted = "\n".join(sorted(formatted))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 87: ordinal not in range(128)
```

Django tries to output messages in `django.core.management.base`, also using `force_str`, so each item in `formatted` will be a utf-8 encoded `str`. Even if it didn't, tho, `CheckMessage` has the decorator `python_2_unicode_compatible`, which overwrites its `__str__` to return a utf-8 encoded `str` already.

But `base` also has `unicode_literal` imported, so the `"\n"` in `"\n".join(sorted(formatted))` is `unicode` and adding to it will attempt to decode every `str` in the iterable `.join()` encounters. Unfortunately it uses the default encoding for this, which is `ascii`.

Googling reveals quite a few strange things about that. In short: there's `sys.setdefaultencoding()`, but Python's own `site.py` deletes it early on.
A work-around is this:

``` python
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
```

And then indeed it works. The problem is that I'm not at all sure that this won't break something else, which might rely on the other default encoding, so I think it's enough to just hide the non-ASCII chars from our check messages.

mbertheau replied:
Yeah, good investigative work! Time to move to Python 3. Or Lisp? :)

mbertheau replied:
`.encode('ascii', 'replace')` :)

mbertheau commented on migrate_config.py:
```diff
@@ -0,0 +1,44 @@
+                    name.replace('"', '\"'),
+                    model.replace('"', '\"'))
+    print '    },'
+print "}"
```
Shouldn't we rather read that from the database? The json file can easily be out of date, the database is guaranteed to be up-to-date (if not necessarily correct; but also not necessarily less so than the json file).

or replied:
It will have to be done on the production database before any migration, so I intend to dump `Event` and `EventData` as fixtures from the production database.

mbertheau commented on migrate_config.py:
```diff
@@ -0,0 +1,44 @@
+event_config = {}
+
+for unused_event_id, event_obj in events.items():
+    print event_obj
```
debug print

mbertheau commented on migrate_config.py:
```diff
@@ -0,0 +1,44 @@
+print "EMARSYS_EVENTS = {"
+for event_name, params in sorted(event_config.items(),
+                                 lambda a, b: cmp(a[0], b[0])):
+    print u'    "{}": {{'.format(event_name.replace('"', '\"'))
```
I think `repr` is meant for that:

``` python
>>> print repr("Ich sagte: \"Du sagtest: 'Nein!'\"")
'Ich sagte: "Du sagtest: \'Nein!\'"'
```

or replied:
``` python
In [1]: print repr("Ich sagte: \"Du sagtest: 'Öh, doch!'\"")
'Ich sagte: "Du sagtest: \'\xc3\x96h, doch!\'"'

In [2]: print '"{}"'.format("Ich sagte: \"Du sagtest: 'Öh, doch!'\"".replace('"', '\\"'))
"Ich sagte: \"Du sagtest: 'Öh, doch!'\""
```

Of course I considered `repr()`, but I didn't want the utf-8 byte representation there. Also keep in mind that this code won't even make it into the end product. It's just a tool to generate the settings.

But this made me realize that `'\"'` should have been `'\\"'`.

mbertheau replied:
Yeah. I was going to suggest putting the code into a gist and linking to the gist in the commit message and/or the README.

mbertheau commented on django_emarsys/context_provider_registry.py:
```diff
@@ -51,3 +51,9 @@ def get_context_provider(event_name):
+def get_placeholder_data(event_name, **kwargs):
+    context_provider = get_context_provider(event_name)
+
+    return context_provider(**kwargs)
```
Completely agree with moving that away from models.

mbertheau commented on tests/trigger_event_tests.py:
```diff
@@ -0,0 +1,116 @@
+    def test_trigger_event_with_unknown_name(self, mock_api_get_events):
+        mock_api_get_events.return_value = {}
+        settings.EMARSYS_EVENTS = {}
+        event = trigger_automatic_event("test event", "oliver.runge@machtfit.de")
```
Not sure you want to expose this poor guy's email address unnecessarily to the relentlessly spamming internets :)

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,70 @@
+from django_emarsys import api
+from django_emarsys.models import NewEvent, NewEventInstance
+
+log = logging.getLogger(__file__)
```
Usually use `getLogger(__name__)`

mbertheau commented:
Concerning the commit message of dfe002dda:

I'd free myself from cyclic dependencies by uncoupling the emarsys logic as much as possible from the persistence layer through putting _no_ code on models and not using a model manager. Indications that this uncoupling might be a good idea are:
- it frees the way to put code where it belongs (the event module)
- it makes it easier for people to implement other persistence backends while keeping the same API
- it's easier to mock in tests, so that those don't hit an actual database

mbertheau commented on django_emarsys/config.py:
```diff
@@ -0,0 +1,184 @@
+    >>>     assert 'DEBUG' in dir(settings)
+    >>>     assert not hasattr(settings, 'DEBUG')
+    >>>
+    >>> test_strange_dir_issue()
```
The part about hasattr is an implementation comment, not documentation, so I'd put it behind `#` comments.

mbertheau commented on django_emarsys/config.py:
```diff
@@ -0,0 +1,184 @@
+                         for event, params
+                         in events.items())
+
+    # logging can happen here if desired
```
rm comment

mbertheau commented on django_emarsys/config.py:
```diff
@@ -0,0 +1,184 @@
+
+    try:
+        apps.get_model(model)
+    except (AttributeError, LookupError, ValueError) as e:
```
Maybe comment why we're catching AttributeError and ValueError.

or replied:
Good idea, I had to try out quite a few things before I found out again why. ;)

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,263 @@
+    except EmarsysError as e:
+        log.error('Emarsys error: {}\n'.format(e))
+        return num_new_events, num_updated_ids, num_deleted_ids, \
+            unknown_event_names, unsynced_event_names
```
Might want to allow the caller to distinguish this failure case from an idempotent but successful sync.

or replied:
Hmmmmm. Any idea about the best way? The correct thing to do would be to re-raise the exception, I guess. But most of the calling code couldn't really make decisions on that anyway.

mbertheau replied:
I say wrap the exception in a django_emarsys exception and raise that. The calling code can then decide to not show a success alert to the user :)

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,263 @@
+                    "configured in settings.EMARSYS_EVENTS: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unknown_event_names)))
```
Not sure if we should warn about these. Events could be used for a lot of stuff and with a lot of apps outside of the app that uses this module, so this could be a common case.

or replied:
Ok. In that case `sync_events` could and probably should ignore all events it isn't configured for, so they never make it into the DB.

mbertheau replied:
Ack.

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,263 @@
+                          data=data,
+                          create_user_if_needed=create_user_if_needed,
+                          contact_data_provider=contact_data_provider,
+                          source=EventInstance.SOURCE_MANUAL)
```
I forgot, why not `trigger_event(..., source=SOURCE_MANUAL)`?

or replied:
`EventInstance.SOURCE_*` shouldn't be exposed, methinks.
It'd also require another import to use the trigger. Either from `django_emarsys.models` or from `django_emarsys.event`.

If you'd prefer a `source` argument (I guess automatic could be the default), then I propose `source="automatic"` or `source="manual"`, which the trigger function could map to `EventInstance`'s constants, but using them directly feels cumbersome.

mbertheau replied:
I agree, they shouldn't be exposed. I think these symbols should not be defined on the model, but rather in a neutral place like `django_emarsys/__init__.py`. The model can use them to persist them, and the API can use them as arguments.

or replied:
I feel they belong to the model, because the model will store those values directly in the database. If the model changes in some way involving the `source` attribute, then all code triggering events breaks, or `event` has to introduce a mapping to handle the new situation, so it's a good idea to have that mapping right away.

I can think of three ways of doing this:
1. `trigger_automatic_event()` and `trigger_manual_event()`
2. `trigger_event(..., source="automatic")` and `trigger_event(..., source="manual")`
3. `trigger_event(..., source=event.AUTOMATIC)` and `trigger_event(..., source=event.MANUAL)` (or other constants in `__init__.py` that are _not_ the same as in the model)

I'm least happy with 3., because of the extra import and knowledge necessary. I think 2. is sufficiently Pythonic, but of course has the typo factor. 1. is (in my eyes) best for the calling code and only has the disadvantage of those 2 extra functions with repeated argument lists.

That being said, I don't _mind_ any of the solutions, so if you feel strongly in favour of either of those, then I'll happily implement that'un. :)

mbertheau replied:
Lately I feel strongly about taking everything away from the models :) that isn't directly related to their task, which is persistence. The models are the persistence layer. The persistence layer doesn't determine which possible sources there are, the business concept we're implementing is. For that reason I don't see "a way the model could change involving the source attribute" (quoted for sponginess ;)) that isn't triggered by a change in the business concept. Models entmachten! Views entmachten! Independent business logic for president!

So I'm for 3., and for pragmatic reasons I don't see why the persistence layer shouldn't use the constants from the business concept directly.

I don't insist on this change, but I hope that my arguments are convincing :)

or replied:
I fully agree on the first two sentence. The difference is, that I think those constants are a vital part of the model's decision how to persist things. Because `source` isn't a plain `CharField` but a `CharField` with choices, which already maps user input to its representation for the persistence layer.

The very fact that the model would have to import these constants from elsewhere in order to even _define_ its `source` field is a strong indicator that it belongs to the model.

We're talking about two different points, tho. One is the location of those constants, the other the exposed functions, because it's conceivable to define the constants elsewhere and STILL expose `trigger_automatic_event`.

I want to propose another way, because `source` is (and likely will remain) a binary argument at heart:
4. `trigger_event(..., manual=True)`

or replied:
This is not an appeal to authority, but https://docs.djangoproject.com/en/1.8/ref/models/fields/#choices recommends:

> Generally, it’s best to define choices inside a model class, and to define a suitably-named constant for each value:

mbertheau replied:
I agree with everything you say. The bool solution sidesteps the discussion, so let's go with that and discuss the other issue at another opportunity :)

mbertheau replied:
Yeah, I'll cite:

> [...] makes the choices easy to reference (e.g, Student.SOPHOMORE will work anywhere that the Student model has been imported)

That's a contradictio in adiectio in my opinion, because "anywhere the model has been imported" is, like, the opposite!, like, of "easy to reference" ^^

So the actual business constants should be defined outside the model. How the model uses these constants for its task is a debatable issue, and also one that is now irrelevant to this PR, since we have the _bool solution_.

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,263 @@
+            recipient_email=recipient_email,
+            emarsys_event_id=emarsys_event_id,
+            data=data,
+            source=source)
```
You could deduplicate that function call with a function-local function

``` python
def _trigger_event(...):
    def doit():
        return _create_event_instance(...)
    ...
    event = doit()
    if ...:
        ...
        event = doit()
    ....
```

mbertheau replied:
Hmm, `_create_event_instance` is a pretty expensive function, and we're calling it twice in some occasions, and in those occasions most of the expensive work doesn't have to be done twice (we just need to repeat the API call).

or replied:
We'd still trigger the event twice, once with an error message. I think it makes sense to track those `EventInstance`s.

mbertheau replied:
Ah, there's more to it, I overlooked that. But the parameter checking is done twice unnecessarily.

This is not critical by any means, I just mention all my thoughts.

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,263 @@
+
+    """
+    if not source:
+        source = EventInstance.SOURCE_AUTOMATIC
```
Can put that in the function args directly as the default, no?

mbertheau commented on django_emarsys/migrations/0003_newevent.py:
```diff
@@ -0,0 +1,27 @@
+            },
+            bases=(models.Model,),
+        ),
+    ]
```
Could you merge migrations such that we have as little new migrations as possible? I believe that makes migrating faster.

or replied:
Yes, but I think the data migration should stay on its own to recover form potential problems.
That'd mean three migrations: preparation, data migration, cleanup. At least better than 5, aye?

mbertheau replied:
Definitely better :).

Do you see potential problems that are not undiscovered bugs in django_emarsys?

or replied:
No, but it just feels like a bad idea to do it all in one.

Just checked whether there's consensus about this question in general:
https://docs.djangoproject.com/en/1.8/topics/migrations/#data-migrations

> Migrations that alter data are usually called “data migrations”; they’re best written as separate migrations, sitting alongside your schema migrations.

Oh, as a more concrete point here: I didn't migrate `Event`s and had no way to migrate `EventData`, that means reversing the data migration _will_ fail, if `Event` and `EventData` are not restored based on fixtures.

I know the reverse migrations are not a huge focus, but I'll make a note of this in the migration file anyway.

mbertheau replied:
Ack :)

mbertheau commented on oscar_emarsys/dashboard/emarsys/app.py:
```diff
@@ -26,12 +25,12 @@ def get_urls(self):
                 self.event_data_lookup_view.as_view(),
                 name='emarsys-event-data-lookup'),
-            url(r'^(?P<pk>\d+)/placeholder-data/$',
+            url(r'^(?P<pk>.+)/placeholder-data/$',
```
PK is still an integer, no?

or replied:
Good catch, that's from the time when pk was the name. SOMEONE forgot about it.

mbertheau commented on pytest.ini:
```diff
@@ -0,0 +1,4 @@
+[pytest]
+python_files=tests/*.py
+norecursedirs = __pycache__
```
Does this module work with Python 3? If not we should maybe remove this line, lest people start making assumptions, then accusations and demands.

mbertheau commented on tox.ini:
```diff
@@ -0,0 +1,3 @@
+[flake8]
+exclude = migrations
+#ignore = W503
```
commenting an ignoration ^^

or committed:
Add migrations

Add migrations to first create all new models, then perform a data
migration, and lastly delete the old models and rename the new models.

The data migration is not out of the box reversible, see comments in
0003_migrate_eventinstances.py for more details.

https://machtfit.atlassian.net/browse/MAC-814

or committed:
Remove unused urls and admin module

https://machtfit.atlassian.net/browse/MAC-814

or committed:
Add new models

NewEvent and NewEventInstance will replace Event and EventInstance
eventually.

https://machtfit.atlassian.net/browse/MAC-814

mbertheau commented on requirements.in:
```diff
@@ -1,3 +1,7 @@
+pytest
+pytest-django
+mock
+jsonfield
```
This is missing the corresponding change in requirements.txt.

mbertheau replied:
Would also be great if you could squash the requirements changes with their corresponding commits.

or replied:
Will do!

mbertheau commented on django_emarsys/test_helper.py:
```diff
@@ -1,5 +1,4 @@
 # -*- coding: utf-8 -*-
-
```
Unnecessary whitespace change

mbertheau commented on django_emarsys/event.py:
```diff
@@ -0,0 +1,237 @@
+        event_name=event_name,
+        recipient_email=recipient_email,
+        source=source,
+        emarsys_id=emarsys_event_id)
```
Hmm, this database write probably never sees the light of a database commit: In the success case, `handle_success` is called, which writes again, making the first write unnecessary. In expected error cases, `handle_error` is called, which writes again. In unexpected error cases, the transaction is very likely to be rollbacked.

So work with an unsaved instance instead?

mbertheau replied:
If you decide to follow my reasoning, then that also makes the sending state unnecessary.

or replied:
I'll leave it in for now, because the object would be saved even with an uncaught exception, potentially helping to track down the problem.

mbertheau commented on django_emarsys/test_helper.py:
```diff
@@ -35,12 +34,8 @@ def assertEmarsysEvents(self, events, all=True):
-            instances = EventInstance.objects.filter(event__name=event_name)
-
-            if user:
-                instances = instances.filter(user_id=user.pk)
```
`user` is unused now, so it should be removed completely.

or replied:
Good point, will do.

mbertheau commented:
It might be easier for other users (if there are any!) to extract `migrate_config.py` if it were a gist that's linked in the changelog.

or commented:
Gists can be deleted and any user finding a gist link in a commit message would also find the script, this way the repository is all she needs.

mbertheau committed:
Update .gitignore, add tox.ini

mbertheau committed:
Add EMARSYS_* configuration and validation

A Django check is implemented to make sure the config is always sanity
checked before the app is used.

https://machtfit.atlassian.net/browse/MAC-814

mbertheau committed:
Make views work with NewEvent and NewEventInstance

Refactors the views to use the new models instead of the deprecated
ones.

https://machtfit.atlassian.net/browse/MAC-814

mbertheau committed:
Remove old models

mbertheau committed:
Rename NewEvent and NewEventInstance

https://machtfit.atlassian.net/browse/MAC-814

mbertheau committed:
Version 0.23

mbertheau committed:
Use explicit relative imports

Less tokens, less duplication.

See also https://mail.python.org/pipermail/python-dev/2013-July/127404.html
and the whole thread.

mbertheau committed:
Move code into event module

This is the module exposed to other apps, providing the necessary
functions to sync events and trigger them.

https://machtfit.atlassian.net/browse/MAC-814
